lib/Dialect/SPIRV/Transforms/DecorateWithLayoutInfoSPIRVPass.cpp

```diff
@@ -0,0 +1,282 @@
+    llvm_unreachable("unhandled SPIR-V composite type");
+  }
+}
+
+unsigned SPIRVLayoutInfoUtils::getPrimitiveTypeSizeInBytes(Type primitiveType) {
+  // According to the SPIR-V spec:
+  // "The ArrayStride, MatrixStride, and Offset decorations must be large
+  // enough to hold the size of the objects they affect (that is, specifying
+  // overlap is invalid)."
+  auto bitWidth = primitiveType.getIntOrFloatBitWidth();
+  if (bitWidth == 1 || bitWidth == 8 || bitWidth == 16)
+    return 4;
+  return bitWidth / 8;
+}
+
+bool SPIRVLayoutInfoUtils::isLegalType(Type type) {
+  auto ptrType = type.dyn_cast<spirv::PointerType>();
+  if (!ptrType) {
+    return true;
+  }
+  auto storageClass = ptrType.getStorageClass();
+  auto structType = ptrType.getPointeeType().dyn_cast<spirv::StructType>();
+  if (!structType) {
+    return true;
+  }
+  if (storageClass != spirv::StorageClass::Uniform &&
+           storageClass != spirv::StorageClass::StorageBuffer &&
+           storageClass != spirv::StorageClass::PushConstant) {
+    return true;
+  }
+  return structType.hasLayout();
+}
+
+bool SPIRVLayoutInfoUtils::isPrimitiveType(Type primitiveType) {
+  if (auto integerType = primitiveType.dyn_cast<IntegerType>()) {
+    return llvm::is_contained(llvm::ArrayRef<unsigned>({1, 8, 16, 32, 64}),
+                              integerType.getWidth());
+  }
+  return primitiveType.isF32() || primitiveType.isF64();
+}
+
+namespace {
+class SPIRVGlobalVariableOpLayoutInfoDecoration
+    : public OpRewritePattern<spirv::GlobalVariableOp> {
+public:
+  using OpRewritePattern<spirv::GlobalVariableOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(spirv::GlobalVariableOp op,
+                                     PatternRewriter &rewriter) const override {
+    // Initial offset is zero.
+    spirv::StructType::LayoutInfo baseOffset = 0;
+    auto ptrType = op.type().cast<spirv::PointerType>();
+    auto structType = SPIRVLayoutInfoUtils::decorateType(
+        ptrType.getPointeeType().cast<spirv::StructType>(), baseOffset);
+    auto decoratedType =
+        spirv::PointerType::get(structType, ptrType.getStorageClass());
+    // Save all named attributes except "type" attribute.
+    SmallVector<NamedAttribute, 4> globalVarAttrs;
+
+    for (const auto &attr : op.getAttrs()) {
+      if (attr.first == "type") {
+        continue;
+      }
+      globalVarAttrs.push_back(attr);
+    }
+
+    rewriter.replaceOpWithNewOp<spirv::GlobalVariableOp>(
+        op, rewriter.getTypeAttr(decoratedType), globalVarAttrs);
+    return matchSuccess();
+  }
+};
+
+class SPIRVAddressOfOpLayoutInfoDecoration
+    : public OpRewritePattern<spirv::AddressOfOp> {
+public:
+  using OpRewritePattern<spirv::AddressOfOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(spirv::AddressOfOp op,
+                                     PatternRewriter &rewriter) const override {
+    auto spirvModule = op.getParentOfType<spirv::ModuleOp>();
+    if (!spirvModule) {
```

Comment:
I dont think you need to check this (or this check needs to be part of the verifier of addressOfOp